src/parse-nm.c

```diff
@@ -808,8 +807,21 @@ netplan_parser_load_keyfile(NetplanParser* npp, const char* filename, GError** e
         _kf_clear_key(kf, "ipv4", "method");
     g_free(tmp_str);
 
-    /* Vlan: XXX: find a way to parse the "link:" (parent) connection */
-    keyfile_handle_generic_uint(kf, "vlan", "id", &nd->vlan_id, G_MAXUINT);
+    /* Handling VLANs */
+    if (nd_type == NETPLAN_DEF_TYPE_VLAN) {
+        keyfile_handle_generic_uint(kf, "vlan", "id", &nd->vlan_id, G_MAXUINT);
+        g_autofree gchar* parent = g_key_file_get_string(kf, "vlan", "parent", NULL);
+
+        if (parent) {
+            /*
+             * Generate a placeholder interface to be the VLAN's parent.
+             * It's required because Network Manager allows the creation of
+             * VLAN connections with non-existing parent interfaces.
+             */
+            nd->vlan_link = netplan_netdef_new(npp, parent, NETPLAN_DEF_TYPE_NM_PLACEHOLDER_, NETPLAN_BACKEND_NM);
```

Comment:
I didn't realize it could be the connection UUID as well hmm...

And it seems to be a bit more complicated than that. For example, if `eth0` was configured via Netplan, it's netdef ID will be `eth0` and it will not have a UUID in the generated key file. Although you can create a VLAN pointing the parent to the NM generated UUID. In this case, we will never find the parent in the netplan state. The same problem will happen if `eth0` was configured using `match` and its netdef ID is not its name, we will not find it by the name `eth0` either.

And there is also the case the parent wasn't parsed yet by Netplan because it's in a different file.

As I see, it's OK to just use whatever parent name was used, even if we already have the parent parsed with a different netdef ID. Network Manager will still be able to create the interface.

What do you think?

Reply:
I added a test.

Regarding the interface name length, for virtual interfaces we don't write the `connection.interface-name` if it's longer than IF_NAMESIZE. But as you said we don't render the placeholder interface. We also have a validation step checking if the name is longer thanbut it's only a warning at the moment. Maybe we should look at this validation again...